app/abtests/Test.scala

```diff
@@ -2,26 +2,42 @@ package abtests
 
 import com.github.slugify.Slugify
 import play.api.libs.json.{JsValue, Json, Writes}
-import play.api.mvc.Cookie
+import play.api.mvc.{Cookie, Request}
+
+import scala.util.matching.Regex
 
 case class Percentage(value: Double) {
   def of[B](x: B)(implicit numB: Numeric[B]): Double = numB.toDouble(x) / 100 * value
 }
 
-case class Test(name: String, audienceSize: Percentage, audienceOffset: Percentage, variants: Variant*) {
+sealed trait Test {
+  val name: String
+  val audienceSize: Percentage
+  val audienceOffset: Percentage
+  val variants: Iterable[Variant]
+
   val startId: Int = audienceOffset.of(Test.maxTestId).toInt
   val endId: Int = startId + audienceSize.of(Test.maxTestId).toInt
   val idRange: Range = startId.to(endId).tail
 
   val slug: String = Test.slugify(name)
   val cookieName: String = s"${Test.cookiePrefix}.${slug}"
 
-  def allocate(id: Int): Option[Allocation] = {
-    if (!idRange.contains(id)) None
-    else Some(Allocation(this, variants.toList(id % variants.size)))
+  def canRun(req: Request[_]): Boolean
+
+  def allocate(id: Int, request: Request[_]): Option[Allocation] = {
+    if (idRange.contains(id) && canRun(request)) Some(Allocation(this, variants.toList(id % variants.size)))
+    else None
   }
 }
 
+case class SplitTest(name: String, audienceSize: Percentage, audienceOffset: Percentage, variants: Variant*) extends Test {
```

Comment:
It looks like you can keep the `Test` `case class` and define `canRun` as a function with a default value